File: C++/145.binary-tree-postorder-traversal.cpp
/*
 * @lc app=leetcode.cn id=145 lang=cpp
 *
 * [145] Binary Tree Postorder Traversal
 */

#include<vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// @lc code=start
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;

        if (!root) return res;

        vector<tuple<TreeNode*, bool>> stack;
        stack.push_back(make_tuple(root, false));

        while (stack.size() > 0) {
            TreeNode *cur = get<0>(stack.back());
            bool visited = get<1>(stack.back());
            stack.pop_back();

            if (cur!=NULL) {
                if (visited) 
                    res.push_back(cur->val);
                else {
                    stack.push_back(make_tuple(cur, true));
                    stack.push_back(make_tuple(cur->right, false));
                    stack.push_back(make_tuple(cur->left, false));
                }
            }
        }
        return res;

        // vector<int> res;
        // if (!root) return res;

        // vector<TreeNode*> stack;
        // stack.push_back(root);

        // while (stack.size()>0) {
        //     TreeNode *cur = stack.back();
        //     stack.pop_back();

        //     res.push_back(cur->val);
        //     if (cur->left) stack.push_back(cur->left);
        //     if (cur->right) stack.push_back(cur->right);
        // }
        // reverse(res.begin(), res.end());
        // return res;
    }
};
// @lc code=end

File: Python/145.binary-tree-postorder-traversal.py
#
# @lc app=leetcode id=145 lang=python
#
# [145] Binary Tree Postorder Traversal
#
# https://leetcode.com/problems/binary-tree-postorder-traversal/description/
#
# algorithms
# Hard (50.85%)
# Likes:    1202
# Dislikes: 65
# Total Accepted:    306.4K
# Total Submissions: 599.1K
# Testcase Example:  '[1,null,2,3]'
#
# Given a binary tree, return the postorder traversal of its nodes' values.
# 
# Example:
# 
# 
# Input: [1,null,2,3]
# ⁠  1
# ⁠   \
# ⁠    2
# ⁠   /
# ⁠  3
# 
# Output: [3,2,1]
# 
# 
# Follow up: Recursive solution is trivial, could you do it iteratively?
# 
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        
        ans = []
        stack = [root]
        while stack:
            node = stack.pop()
            ans.append(node.val)
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)
        return ans[::-1]

            
        # res = []
        # stack = [root]
        # while stack:

        #     cur = stack.pop()    
        #     if cur:
        #         res.append(cur.val)
        #         stack.append(cur.left)        
        #         stack.append(cur.right)

        # return res[::-1]
# @lc code=end

sol = Solution()
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
print(sol.postorderTraversal(root))